packages/gatsby/src/schema/query-loki.js

```diff
@@ -0,0 +1,194 @@
+// (accumulator) must be a reference to an empty object. The converted
+// fields will be added to it. E.g
+//
+// {
+//   internal: {
+//     type: {
+//       $eq: "TestNode"
+//     },
+//     content: {
+//       $regex: new MiniMatch(v)
+//     }
+//   },
+//   id: {
+//     $regex: newMiniMatch(v)
+//   }
+// }
+//
+// After execution, acc would be:
+//
+// {
+//   "internal.type": {
+//     $eq: "TestNode"
+//   },
+//   "internal.content": {
+//     $regex: new MiniMatch(v)
+//   },
+//   "id": {
+//     $regex: // as above
+//   }
+// }
+function dotNestedFields(acc, o, path = ``) {
+  if (_.isPlainObject(o)) {
+    if (_.isPlainObject(_.sample(o))) {
+      _.forEach(o, (v, k) => {
+        dotNestedFields(acc, v, path + `.` + k)
+      })
+    } else {
+      acc[_.trimStart(path, `.`)] = o
+    }
+  }
+}
+
+// Converts graphQL args to a loki query
+function convertArgs(gqlArgs) {
+  const dottedFields = {}
+  dotNestedFields(dottedFields, toMongoArgs(gqlArgs.filter))
+  return dottedFields
+}
+
+// Converts graphql Sort args into the form expected by loki, which is
+// a vector where the first value is a field name, and the second is a
+// boolean `isDesc`. Nested fields delimited by `___` are replaced by
+// periods. E.g
+//
+// {
+//   fields: [ `frontmatter___date`, `id` ],
+//   order: `desc`
+// }
+//
+// would return
+//
+// [ [ `frontmatter.date`, true ], [ `id`, true ] ]
+function toSortFields(sortArgs) {
+  const { fields, order } = sortArgs
+  return _.map(fields, field => [
+    field.replace(/___/g, `.`),
+    _.lowerCase(order) === `desc`,
+  ])
+}
+
+// Ensure there is an index for each query field. If the index already
+// exists, this is a noop (handled by lokijs).
+function ensureIndexes(coll, findArgs) {
+  _.forEach(findArgs, (v, fieldName) => {
+    coll.ensureIndex(fieldName)
+  })
+}
+
+/**
+ * Runs the graphql query over the loki nodes db.
+ *
+ * @param {Object} args. Object with:
+ *
+ * {Object} gqlType: built during `./build-node-types.js`
+ *
+ * {Object} rawGqlArgs: The raw graphql query as a js object. E.g `{
+ * filter: { fields { slug: { eq: "/somepath" } } } }`
+ *
+ * {Object} context: The context from the QueryJob
+ *
+ * {boolean} firstOnly: Whether to return the first found match, or
+ * all matching result.
+ *
+ * @returns {promise} A promise that will eventually be resolved with
+ * a collection of matching objects (even if `firstOnly` is true)
+ */
+function runQuery({ gqlType, rawGqlArgs, context = {}, firstOnly }) {
+  // Clone args as for some reason graphql-js removes the constructor
+  // from nested objects which breaks a check in sift.js.
+  const gqlArgs = JSON.parse(JSON.stringify(rawGqlArgs))
+
+  const lokiArgs = convertArgs(gqlArgs)
+
+  const coll = getDb().getCollection(gqlType.name)
+
+  // Allow page creators to specify that they want indexes
+  // automatically created for their pages.
+  if (context.useQueryIndex) {
+    ensureIndexes(coll, lokiArgs)
```

Comment:
How expensive this is when indexes were already applied? We definitely will run same query (with different context) when we use same template for multiple pages

This has also potential to keep unneeded indexes - for example when we modify filters (remove some fields from filter), but not sure if it's worth keeping track of that

Reply:
@pieh loki handles the case of the index already existing so there's no performance penalty.

But yes, this will keep old indexes around if filters are modified. My solution here for applying queries is very basic, but I think there's a lot more we could do here over time (see original PR description). My suggestion would be to keep as is, but iterate on this based on feedback